### DISCUSSION TITLE
Question on pinocchio::computeCoriolisMatrix

### DISCUSSION THREAD

**Discussion by xiaosuyang** [stack-of-tasks/pinocchio#2298] [Q&A] [answered]
Hello. When I use  ``pinocchio::computeCoriolisMatrix(robot.m_model,data,q,v)``, there is a assertion that:

![image](https://github.com/stack-of-tasks/pinocchio/assets/55594017/e9fd3501-1cb2-4d97-bcc2-2a32bb265899)

It seems that the quaternion of ``q`` has some problem, but there is no problem when I use ``pinocchio::computeGeneralizedGravity(robot.m_model,data,q)``. How can I fix it?

**Comment by xiaosuyang** [comment]
Can I use pinocchio::NonlinearEffect to replace pinocchio::computeCotiolisMatrix and  pinocchio::computeGeneralizedGravity?

**Reply by fabinsch** [reply, marked as answer]
Hi @xiaosuyang, indeed there seems to be a problem with your q. You could use Pinocchio's helper functions to make sure a configuration is valid by
```
q = pin.normalize(model, q)
```
I hope this helps.

**Reply by jcarpent** [reply]
> Can I use pinocchio::NonlinearEffect to replace pinocchio::computeCotiolisMatrix and pinocchio::computeGeneralizedGravity?

Yes, this is the best way to proceed to avoid unnecessary computations.
